### PR TITLE
fullsum training automaton: RNA/CTC/HMM topology

### PR DESCRIPTION
**Pull request by ZhouW321**
sprint automaton based  from-scratch full-sum training using BW Loss

### CODE REVIEW

File: src/Nn/AllophoneStateFsaExporter.cc [hunk 1]
```diff
@@ -190,4 +325,60 @@ AllophoneStateFsaExporter::ExportedAutomaton AllophoneStateFsaExporter::exportFs
     return result;
 }
 
+// Note: pronunciation variants ignored here, i.e. no normalization
```
Comment (Marvin84) [inline src/Nn/AllophoneStateFsaExporter.cc:328]: Do I get it right that here you overwrite the arc weights with only transition scores?
If yes, you are ignoring both lemma level and lexicon level pronunciation normalizations. This makes the implementation not generic enough.
And the question is: why do you even need to have the HMM topology modified here? We already have an FSA builder module `ClassicTransducerBuilder` that prepares the FSA for full-sum with our standard HMM topology that supports both mono-state and 3-partite. Shouldn't you extend that?

File: src/Nn/AllophoneStateFsaExporter.cc [hunk 2]
```diff
@@ -135,11 +123,115 @@ static void make_single_final_state(std::vector<Fsa::StateId>& states, std::vect
...
+        {
+            Fsa::StateId target = arc->target_;
+            if ( target == sId )
+                continue;
+            u32 depth = getStateDepth(target, stateDepth, fsa) + 1;
+            if ( depth < stateDepth[sId] )
+                stateDepth[sId] = depth;
+        }
+        if ( state->isFinal() )
+            stateDepth[sId] = 0;
```
Comment (SimBe195) [inline src/Nn/AllophoneStateFsaExporter.cc:142]: Why do we need to iterate over the arcs if the state is final? Shouldn't the for loop be only in the "else" case of `if (state->isFinal() )`

File: src/Nn/AllophoneStateFsaExporter.cc [hunk 3]
```diff
@@ -190,4 +325,60 @@ AllophoneStateFsaExporter::ExportedAutomaton AllophoneStateFsaExporter::exportFs
...
+void AllophoneStateFsaExporter::modifyMinDuration(std::vector<Edge>& edges, std::vector<Fsa::StateId>& states, Core::Ref<const Fsa::StaticAutomaton> automaton, f64 time) const
+{   // expanded sequence should not exceed number of frames
+    u32 expand = minOccur_;
+    if ( time >= 0 ) 
+    {   // if negative time: unknown length, assume always expandable
+        u32 maxLength = time / frameShift_ / reduceFrameFactor_;
+        std::vector<u32> stateDepth(automaton->size(), Core::Type<u32>::max);
+        u32 seqLength = getStateDepth(0, stateDepth, automaton); // shortest sequence length
+        while ( seqLength * expand > maxLength && expand > 1 )
+            --expand; // reduce expansion
```
Comment (SimBe195) [inline src/Nn/AllophoneStateFsaExporter.cc:358]: I would say it's unexpected behavior that the minimum duration may be reduced automatically. In my opinion the user should receive a warning when that happens.

**Review by SimBe195 — CHANGES_REQUESTED** [review]

**Comment by ZhouW321** [comment]
no need anymore

**Pull request closed without merging**